### DISCUSSION TITLE
Can't run locally

### DISCUSSION THREAD

**Discussion by snirego** [dubinc/dub#451] [Q&A] [answered]
Hey,

First of all thank you for this great community and useful open source project.

So I added all the required .env variables and I am trying to run it locally for few days now. I keep getting error messages when I try to build the project so for now I am stuck on this error:
web:build: cache miss, executing 9093f012917e360d
web:build:
web:build: > web@ build MY_LONG_PATH_TO_DIRECTORY\dub\apps\web
web:build: > prisma generate & next build
web:build:
web:build: Environment variables loaded from .env
web:build: Prisma schema loaded from prisma\schema.prisma
web:build: 
web:build: ✔ Generated Prisma Client (v5.2.0) to .\..\..\node_modules\.pnpm\@prisma+client@5.2.0_prisma@5.2.0\node_modules\@prisma\client in 370ms
web:build: Start using Prisma Client in Node.js (See: https://pris.ly/d/client)
web:build: ```
web:build: import { PrismaClient } from '@prisma/client'
web:build: const prisma = new PrismaClient()
web:build: ```
web:build: or start using Prisma Client at the edge (See: https://pris.ly/d/accelerate)
web:build: ```
web:build: import { PrismaClient } from '@prisma/client/edge'
web:build: const prisma = new PrismaClient()
web:build: ```
web:build: 
web:build: See other ways of importing Prisma Client: http://pris.ly/d/importing-client
web:build: 
web:build: unhandledRejection ReferenceError: Headers is not defined
web:build:     at MY_LONG_PATH_TO_DIRECTORY\dub\node_modules\.pnpm\next@14.0.0_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\compiled\next-server\app-route.runtime.prod.js:14:8846
web:build:     at MY_LONG_PATH_TO_DIRECTORY\dub\node_modules\.pnpm\next@14.0.0_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\compiled\next-server\app-route.runtime.prod.js:14:41308
web:build:     at Object.<anonymous> (MY_LONG_PATH_TO_DIRECTORY\dub\node_modules\.pnpm\next@14.0.0_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\compiled\next-server\app-route.runtime.prod.js:14:41329)
web:build:     at Module._compile (node:internal/modules/cjs/loader:1103:14)
web:build:     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
web:build:     at Module.load (node:internal/modules/cjs/loader:981:32)
web:build:     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
web:build:     at Module.require (node:internal/modules/cjs/loader:1005:19)
web:build:     at Module.mod.require (MY_LONG_PATH_TO_DIRECTORY\dub\node_modules\.pnpm\next@14.0.0_@babel+core@7.17.8_@opentelemetry+api@1.4.1_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\server\require-hook.js:64:28)
web:build:     at require (node:internal/modules/cjs/helpers:102:18)
web:build:  ELIFECYCLE  Command failed with exit code 1.
web:build: ERROR: command finished with error: command (MY_LONG_PATH_TO_DIRECTORY\dub\apps\web) pnpm run build exited (1)
web#build: command (MY_LONG_PATH_TO_DIRECTORY\dub\apps\web) pnpm run build exited (1)

 Tasks:    2 successful, 3 total
Cached:    0 cached, 3 total
  Time:    25.717s
Failed:    web#build

 ERROR  run failed: command  exited (1)
 ELIFECYCLE  Command failed with exit code 1.

Is anybody knows what to do next?

**Comment by tomy0000000** [comment]
I ran these: 

```shell
turbo build --filter='@dub/tailwind-config'
turbo build --filter='@dub/ui'
turbo build --filter='@dub/utils'
turbo build
```

Then finally

```
pnpm dev
```

**Comment by versecafe** [comment]
@snirego @snirego Is this still an issue you're experiencing?

**Comment by jaiswalshash** [comment]
Hey, I ran the turbo commands and then ran `pnpm dev` but still I am getting this error

`@dub/utils:dev: DTS Build start
web:dev: [2]     React Email 2.0.0
web:dev: [2] - Getting react-email preview server ready...
web:dev: [2]
web:dev: [2]     Running preview at:          http://localhost:3333
web:dev: [2]
web:dev: [1] Environment variables loaded from .env
web:dev: [1] Prisma schema loaded from prisma\schema
web:dev: [0] Environment variables loaded from .env
web:dev: [0] Prisma schema loaded from prisma\schema
web:dev: [2] Error: > Couldn't find any `pages` or `app` directory. Please create one under the project root
web:dev: [2]     at findPagesDir (C:\Users\jaisw\Desktop\Dub.co\dub\node_modules\.pnpm\next@14.0.5-canary.46_@babel+core@7.23.0_@opentelemetry+api@1.8.0_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\lib\find-pages-dir.js:42:15)
web:dev: [2]     at initialize (C:\Users\jaisw\Desktop\Dub.co\dub\node_modules\.pnpm\next@14.0.5-canary.46_@babel+core@7.23.0_@opentelemetry+api@1.8.0_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\server\lib\router-server.js:71:69)
web:dev: [2]     at async NextCustomServer.prepare (C:\Users\jaisw\Desktop\Dub.co\dub\node_modules\.pnpm\next@14.0.5-canary.46_@babel+core@7.23.0_@opentelemetry+api@1.8.0_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\server\next.js:240:28)
web:dev: [2] 
web:dev: [2] shutting down dev server
@dub/ui:dev: Browserslist: caniuse-lite is outdated. Please run:
@dub/ui:dev:   npx update-browserslist-db@latest
@dub/ui:dev:   Why you should do it regularly: https://github.com/browserslist/update-db#readme
web:dev: [2] email dev --port 3333 exited with code 1
web:dev: --> Sending SIGTERM to other processes..
web:dev: [1] prisma generate && prisma studio --browser none exited with code 1
web:dev: --> Sending SIGTERM to other processes..
web:dev: [0] prisma generate && next dev --turbo --port 8888 exited with code 1
web:dev:  ELIFECYCLE  Command failed with exit code 1.
web:dev: ERROR: command finished with error: command (C:\Users\jaisw\Desktop\Dub.co\dub\apps\web) C:\Users\jaisw\AppData\Roaming\npm\pnpm.cmd run dev exited (1)
web#dev: command (C:\Users\jaisw\Desktop\Dub.co\dub\apps\web) C:\Users\jaisw\AppData\Roaming\npm\pnpm.cmd run dev exited (1)

 Tasks:    0 successful, 3 total
Cached:    0 cached, 3 total
  Time:    9.182s
Failed:    web#dev

 ERROR  run failed: command  exited (1)`
 
 I am executing this command in root and there is the directory named app but it's still showing this error, how can I fix this? 
 Thank You

**Comment by AatirNadim** [comment, marked as answer]
hey @snirego @jaiswalshash  and any other party concerned, who is facing the issue of getting the application up locally

1. Please make sure you are following all the steps present here [dub-local-development](https://dub.co/docs/local-development).
2. Also ensure that you are populating all the required env vars in the `apps/web` env file as well.
3. Please make sure you are using `pnpm` and the right version as  well `"packageManager": "pnpm@8.6.10"`, as present in the root `package,json` file.
4. If you are using a different version of pnpm in general then you can run `pnpx pnpm@8.6.10 i`, in the root directory, i.e., where the root `package.json` file is present, and subsequently, ```pnpx pnpm@8.6.10 dev```. 

```running the dev script will have turbo handle all the js code generation, you dont have to explicitly run turbo commands```
